Allows for gyms from optional subregions to be selected for the Defeat Gym Leader quest

rocketguy2 opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Please fill out this form, so that we have all the info needed to review your changes -->
<!-- Any text inside of the angle brackets will not show up in the final comment. Check the Preview tab to confirm -->

Whenever the DefeatGymQuest is ran, it now has a chance to give a gym from the Orange Islands, Orre, or Magikarp Jump.

## Description
<!-- Describe your changes in detail -->
<!-- If you are adding new content, please let us know if it is based on some canon, and provide a reference to it -->

The DefeatGymQuest now has a chance, equal to selecting a region, to pick a gym from a subregion. At this point, it then chooses from the three subregions with gyms currently in the game, Orange Islands, Orre, and Magikarp Jump. It then runs as normal, using the gyms from the subregion selected.

## Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- Please also link to any related issues or PRs here. -->
<!-- You can link an issue to be auto-closed when this is merged by writing 'Closes #1234' or 'Fixes #1234' -->

This is an implementation of, and therefore Closes #6101 

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

I have generated a quest that picks a gym from each of the subregions on a savefile that has all of the subregions unlocked, which all worked properly.

## Types of changes
<!-- What types of changes does your code introduce? Delete any that don't apply, and/or add more you think would be useful -->

- Expansion of pre-existing feature

rocketguy2 committed:
Allows for gyms from optional subregions to be selected for the Defeat Gym Leader quest

rocketguy2 committed:
Update translations

RegisCoaxans reviewed (changes requested):
The subregion are hard coded, and the way it picks them rather than main gyms is uncanny.
You could instead retrieve these with something like `GameConstants.regionGyms.slice(GameConstants.Region.final)`, and filter out these whose first gym isn't unlocked and already cleared. I don't mind if subregions are as likely to be picked as main regions.

rocketguy2 committed:
Rewrote code to select the gym region in order to prevent hardcoding

Additionally, this code resolves the conflict with PR #5665, by including the fix it made.

rocketguy2 committed:
Merge branch 'develop' into OptionalRegionGymQuest

RegisCoaxans committed:
Fix region filter

rocketguy2 committed:
Merge remote-tracking branch 'upstream/develop' into OptionalRegionGymQuest

rocketguy2 committed:
Merge branch 'develop' into OptionalRegionGymQuest
